tests/delete/models.py

```diff
@@ -147,3 +159,49 @@ class SecondReferrer(models.Model):
     other_referrer = models.ForeignKey(
         Referrer, models.CASCADE, to_field='unique_field', related_name='+'
     )
+
+
+class DeleteTop(models.Model):
+    b1s = GenericRelation('GenericB1')
+    b2s = GenericRelation('GenericB2')
+
+
+class B1(models.Model):
+    delete_top = models.ForeignKey(DeleteTop, models.CASCADE)
+
+
+class B2(models.Model):
+    delete_top = models.ForeignKey(DeleteTop, models.CASCADE)
+
+
+class DeleteBottom(models.Model):
+    b1 = models.ForeignKey(B1, models.RESTRICT)
+    b2 = models.ForeignKey(B2, models.CASCADE)
+
+
+class GenericB1(models.Model):
+    content_type = models.ForeignKey(ContentType, on_delete=models.CASCADE)
+    object_id = models.PositiveIntegerField()
+    generic_delete_top = GenericForeignKey('content_type', 'object_id')
+
+
+class GenericB2(models.Model):
+    content_type = models.ForeignKey(ContentType, on_delete=models.CASCADE)
+    object_id = models.PositiveIntegerField()
+    generic_delete_top = GenericForeignKey('content_type', 'object_id')
+    generic_delete_bottoms = GenericRelation('GenericDeleteBottom')
+
+
+class GenericDeleteBottom(models.Model):
+    generic_b1 = models.ForeignKey(GenericB1, models.RESTRICT)
+    content_type = models.ForeignKey(ContentType, on_delete=models.CASCADE)
+    object_id = models.PositiveIntegerField()
+    generic_b2 = GenericForeignKey()
+
+
+def callback(sender, **kwargs):
+    pass
+
+
+# Prevent collector fast deletes by attaching a signal.
+pre_delete.connect(callback, sender=GenericDeleteBottom)
```

Comment:
This seems unnecessary.

Reply:
This is here to ensure fast deletes don't work for this model so that [these lines](https://github.com/django/django/pull/7364/files#r331456223) are covered by tests.

If this signal is removed, `test_restrict_generic_foreign_key_and_no_fast_delete` will pass even if we remove those two lines.

Reply:
We need [`can_fast_delete()`](https://github.com/django/django/blob/cabf1fe37e46b00c14ed025a176628e748837658/django/db/models/deletion.py#L126) to return `False` for those lines to be covered and this is one way to do it. I can also try to achieve this by setting up a foreign key the right way instead of using this signal, though, if you think that's better.

Reply:
Yes we should test a real use case instead of emulating this path.

Reply:
Changed this to be based on a related foreign key instead of a signal